pkg/kubelet/cm/cpumanager/cpu_assignment.go

```diff
@@ -518,6 +546,51 @@ func (a *cpuAccumulator) takeFullSockets() {
 		a.take(cpusInSocket)
 	}
 }
+func (a *cpuAccumulator) takeFullUnCore() {
+	for _, uncore := range a.freeUncoreCache() {
+		cpusInUncore := a.topo.CPUDetails.CPUsInUncoreCaches(uncore)
+		if !a.needsAtLeast(cpusInUncore.Size()) {
+			continue
+		}
+		a.take(cpusInUncore)
+	}
+}
+
+// First try to take full UncoreCache, if available and need is at least the size of the UncoreCache group.
+// Second try to take the partial UncoreCache if available and the request size can fit w/in the UncoreCache.
+func (a *cpuAccumulator) takeUncoreCache() {
+	for _, uncore := range a.sortAvailableUncoreCaches() {
+		// take full UncoreCache if the CPUs needed is greater than free UncoreCache size
+		if a.numCPUsNeeded >= a.topo.NumCPUs/a.topo.NumUncoreCache {
+			a.takeFullUnCore()
+		}
+
+		if a.isSatisfied() {
+			return
+		}
+
+		// if SMT is enabled, number of cores needed is half the number of CPUs needed
+		// if SMT is disabled, number of cores needed is equal to number of CPUs
+		numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore()
+
+		var freeCPUsInUncoreCache cpuset.CPUSet
+		// when CPUs needed is smaller than a free UncoreCache size, try to assign cores from a single UncoreCache
+		freeCoresInUncoreCache := a.details.CoresNeededInUncoreCache(numCoresNeeded, uncore)
+		klog.V(4).InfoS("free cores from a.details list: ", "freeCoresInUncorecache", freeCoresInUncoreCache)
+		for _, coreID := range freeCoresInUncoreCache.List() {
+			freeCPUsInUncoreCache = freeCPUsInUncoreCache.Union(a.topo.CPUDetails.CPUsInCores(coreID))
```

Comment:
and here we go back to cpus aka pohysical cpus. We should simplify this code and always reason in terms of cores or cpus. Let's review what other functions do and act accordingly.

Reply:
We could potentially split that out into another helper function